### PR TITLE
PLT add plot with iterations

### PR DESCRIPTION
**Pull request by tomMoral**
Add plot relative to iteration numbers.

Also quick start a documentation page on plotting tools. This should be populated at some point.... :(

### CODE REVIEW

**Comment by mathurinm** [comment]
I got confused by the strong change when switching to iteration curve. Can we rename to "objective per iteration" and "suboptimality per iteration" ?
How do we sell it for solvers based on tolerance ?

|||
|---|---|
|image|image|

**Comment by mathurinm** [comment]
For the interpolation part, 

||main| this PR|
|---|---|---|
|1 rep|image|image|
|5 reps|image|image

File: benchopt/plotting/plot_objective_curve.py [hunk 1]
```diff
@@ -1,18 +1,58 @@
...
+EPS = 1e-10
 
 
 def _remove_prefix(text, prefix):
     return text[len(prefix):] if text.startswith(prefix) else text
 
 
+def _get_curve_interp(df, obj_col, q_min, q_max):
+    """Compute interpolation curve for a given solver.
+
```
Comment (tomMoral) [inline benchopt/plotting/plot_objective_curve.py:20]:
```suggestion
    """Compute interpolation curve for a given solver.

    For each repetition, we compute an interpolation curve on a fixed
    grid and then compute the median and the quartiles to display
    a curve and the error curve.

```

File: benchopt/plotting/plot_objective_curve.py [hunk 2]
```diff
@@ -1,18 +1,58 @@
...
+        The benchmark results filtered for one solver.
+    obj_col : str
+        Column to select in the DataFrame for the plot.
+    q_min, q_max : float
+        The quantile to display the error bars in the plot.
+    """
+    n_interp = min(100, df.groupby('idx_rep').size().max())
+    t = np.logspace(
+        np.log10(df['time'].min()), np.log10(df['time'].max()), n_interp
+    )
```
Comment (tomMoral) [inline benchopt/plotting/plot_objective_curve.py:33]:
```suggestion
    # Compute the size of the grid as the min between 100 points
    # and the number of point in the longest curve. The grid is then
    # chosen in log scale between the min and max time.
    n_interp = min(100, df.groupby('idx_rep').size().max())
    t = np.logspace(
        np.log10(df['time'].min()), np.log10(df['time'].max()), n_interp
    )
```

File: benchopt/plotting/plot_objective_curve.py [hunk 3]
```diff
@@ -1,18 +1,58 @@
...
+    curve_t = (
+        df.groupby('idx_rep').apply(
+            lambda x: pd.DataFrame({
+                't': t,
+                # Linear interpolator to resample on a grid t
```
Comment (tomMoral) [inline benchopt/plotting/plot_objective_curve.py:38]:
```suggestion
                # Linear interpolator to resample the objective curve
                # (x['time'], x[obj_col]) on a grid t
```

File: benchopt/plotting/plot_objective_curve.py [hunk 4]
```diff
@@ -1,18 +1,58 @@
...
+                    bounds_error=False,
+                    fill_value=(
+                        x[obj_col].iloc[0],
+                        x[obj_col].iloc[-1]
+                    )
+                )(t)
+            })
+        )
+    )
+    curve_t = curve_t.groupby('t')['v'].quantile([0.5, q_min, q_max]).unstack()
```
Comment (tomMoral) [inline benchopt/plotting/plot_objective_curve.py:50]:
```suggestion
    # Compute the quartiles on the resampled curves.
    curve_t = curve_t.groupby('t')['v'].quantile([0.5, q_min, q_max]).unstack()
```

File: doc/plotting.rst [hunk 5]
```diff
@@ -0,0 +1,5 @@
+Plotting tools
+==============
+
+
+WIP decsription of the different plotting tools
```
Comment (tomMoral) [inline doc/plotting.rst:5]:
```suggestion
WIP description of the different plotting tools
```

File: doc/whats_new.rst [hunk 6]
```diff
@@ -49,6 +49,14 @@ CLI
 
 - Replace ``-p`` flag by ``-o`` for Objective, by `Mathurin Massias`_ (:gh:`281`).
 
+
+Plotting tools
+~~~~~~~~~~~~~~
+
+- Add plot relative to the number of iterations for research benchmarks,
+  by `Thomas Moreau`_ (:gh:`287`).
+
```
Comment (tomMoral) [inline doc/whats_new.rst:58]:
```suggestion

- Change quantiles for error bar to 0.25 and 0.75,
  by `TomDLT`_ (:gh:`287`).
```

File: doc/whats_new.rst [hunk 7]
```diff
@@ -49,6 +49,16 @@ CLI
...
+- Change quantiles for error bar to 0.25 and 0.75,
+  by `TomDLT`_ (:gh:`287`).
```
Comment (tomMoral) [inline doc/whats_new.rst:60]:
```suggestion
  by `Tom Dupré la Tour`_ (:gh:`287`).
```

File: benchopt/plotting/plot_objective_curve.py [hunk 8]
```diff
@@ -76,28 +128,42 @@ def plot_objective_curve(df, obj_col='objective_value', plotly=False,
...
+                df_.groupby('stop_val')[obj_col]
+                .quantile([q_min, 0.5, q_max]).unstack()
+            )
+        else:
+            curve_t = _get_curve_interp(df_, obj_col, q_min=q_min, q_max=q_max)
 
-        fill_between_x(
-            fig, curve['time'], q1, q9, curve[obj_col], color=CMAP(i % CMAP.N),
-            marker=markers[i % len(markers)], label=solver_name, plotly=plotly
+        fill_between_y(
```
Comment (TomDLT) [inline benchopt/plotting/plot_objective_curve.py:144]: I am not sure I prefer the interpolated plot. I understand it is weird that the curves goes back in time, but it is easily explained by variations in computational time. With the interpolation, the curves look like the solver is not always improving over iterations, and the explanation is more convoluted.

File: benchopt/plotting/plot_objective_curve.py [hunk 9]
```diff
@@ -76,28 +128,42 @@ def plot_objective_curve(df, obj_col='objective_value', plotly=False,
             plt.text(0.5, 0.5, "Not Available")
         return fig
 
+    # use 2nd and 8th decile for now
```
Comment (TomDLT) [inline benchopt/plotting/plot_objective_curve.py:131]: outdated

File: benchopt/plotting/__init__.py [hunk 10]
```diff
@@ -8,6 +8,8 @@
 from .plot_objective_curve import plot_objective_curve  # noqa: F401
 from .plot_objective_curve import plot_suboptimality_curve  # noqa: F401
 from .plot_objective_curve import plot_relative_suboptimality_curve  # noqa: F401 E501
+from .plot_objective_curve import plot_iteration_suboptimality_curve  # noqa: F401 E501
+from .plot_objective_curve import plot_iteration_curve  # noqa: F401
```
[the remark this answers is not in the document]
Comment (tomMoral) [reply benchopt/plotting/__init__.py:12]: I agree we should change this, moving it to a separate file.
The original idea was to be able to easily list all plotting utils without manually adding them but there are better way to do this.

File: benchopt/plotting/plot_objective_curve.py (hunk 9 above)
Comment (tomMoral) [reply benchopt/plotting/plot_objective_curve.py:131]:
```suggestion
    # use 1st and 3rd quartile for error bars
```

**Comment by tomMoral** [comment]
Closing this one in favor of #479

**Pull request closed without merging**